CODING.md

```diff
@@ -0,0 +1,206 @@
+* If that's not possible they should limit themselves to as few managers as possibles
+* If what needs to be tested is the full interaction across many or all managers then we have or can write tests for this in overlord/managers_test.go. Fixture setup in these cases is very costly but they are still easier to iterate on and can be useful to probe behaviour in more internal details than functional/integration tests.
+
+We do not have strong policies against repetition in test code, as usual the important consideration is readability. This area is mostly left to the personal judgement of developers. If any general advice can be given is that:
+
+* investing in clear helpers to setup complex fixtures is often valuable, while compressing actual ad hoc testing and checking code less so, as it might result in if-trees that might be hard to follow
+* wherever applicable, tabular tests (where cases are expressed as a slice of anonymous structs) should be used. For example, they are often appropriate when testing error cases of functions
+
+## Functional/integration tests
+
+We write them using [spread](https://github.com/snapcore/spread). Generally all externally visible features and behaviour should have spread tests. It's also important to have them for robustness and fallback behaviour as related to system properties.
+
+## PRs and refactorings
+
+* PR should ideally have diffs of around 500 lines or less. There might be exceptions when size is due to large repetitive tests, but not for the production code. Experience indicates that smaller PRs are easier to review, while it is hard to do careful and punctual reviews for very large diffs.
+
+* It is fair for reviewers to ask for large PRs to be split. It is also fair to ask for discussion on best strategies to do this with colleagues and architects.
+
+* Whenever reasonable, avoid spurious differences between the code in master and the new code.
+
+* Large mechanical refactoring and changes should be done as separate PRs. Try to separate behaviour changes and refactoring into different PRs and not mix the two.
+
+* Large moving of code around and changes to code placement might also be better done separately.
+
+* PR summaries are expected to be of this form:
+  * *`affected full packages:  short summary in lowercase`*
+    * When too many packages are involved, many can be used instead, or sometimes package names can be abbreviated by using single letters for the top-level package, when non ambiguous combined with the subpackage.
+    * Examples:
+      * `overlord/devicestate: add test to check connect hooks don't break anything`
+      * `gadget,image: remove LayoutConstraints struct`
+      * `o/snapstate: add helpers to get user and gating holds`
+      * `many: correct struct fields and output key`
+  * When no go code is involved, the context prefix before the colon can refer to directories or top-level files instead.
```

Comment:
```suggestion
  * When no Go code is involved, the context prefix before the colon can refer to directories or top-level files instead.
```

Reply:
I used golang as the rest of the document